# Практическая работа №1: Моделирование стандартных дискретных сигналов

Выполнили студенты гр. 0382 Азаров Максим и Шангичев Василий. Вариант №13.

## Цель работы

Изучить математическое описание стандартных дискретных сигналов и овладеть программными средствами их моделирования.

## Основные теоретические положения

__Единичный цифровой импульс__:
$$\delta_d(k) = \begin{cases}
1,~~ k=0 \\
0,~~ k\neq 0
\end{cases} , ~~~~~
k= ~...,-2,-1,0,1,2,...$$

\\
__Дискретный единичный скачок__:
$$\sigma_d(k) = \begin{cases}
1,~~ k \geq 0 \\
0,~~ k < 0
\end{cases} , ~~~~~
k= ~...,-2,-1,0,1,2,...$$

\\
__Дискретная экспоненциальная функция:__:
$$
s_1(k) = \begin{cases}
a^k,~~ k \geq 0 \\
0,~~ k < 0
\end{cases} , ~~~~~
k= ~...,-2,-1,0,1,2,...$$

\\
__Дискретный комплексный гармонический сигнал__:
  $$s_2 ( k ) = C  e^{ j ω_0 k } , ~~~~~
k= ~...,-2,-1,0,1,2,...$$ \\
при разложении по формуле Эйлера, описывается двумя вещественными последовательностями – _косинусоидой_ (вещественная часть) и _синусоидой_
 (мнимая часть)
 $$s_2(k) = C\cos(w_0 k) +j~C\sin(w_0 k)$$

## Постановка задачи

С помощью программных средств провести моделирование и анализ стандартных дискретных последовательностей. Результаты подкрепить соответствующими графиками и выводами.



1. Смоделировать единичный цифровой импульс $\delta_d (k)$ с выводом графиков на интервале дискретного времени $nT \in [0, (N-1)T]$ и дискретного нормированного времени $n \in [0, N-1]$. Пояснить:

$\qquad$a. взаимосвязь между дискретным и дискретным нормированным временем;

$\qquad$b. различие между единичным цифровым импульсом и дельта-функцией.

2. Смоделировать дискретный единичный скачок $\sigma_d (k)$ с выводом графиков на интервале дискретного времени $nT \in [0, (N-1)T]$ и дискретного нормированного времени $n \in [0, N-1]$. Пояснить:

$\qquad$a. соответствие между дискретным и аналоговым единичными скачками;

$\qquad$b. чему равна частота дискретизации дискретного единичного скачка.

3. Смоделировать дискретную экспоненциальную функцию $s_1 (k)$ с выводом графиков на интервале дискретного времени $nT \in [0, (N-1)T]$ и дискретного нормированного времени $n \in [0, N-1]$. Пояснить соответствие между дискретной и аналоговой экспонентами.


4. Смоделировать дискретный комплексный гармонический сигнал $s_2 (k) = C exp(j \hat\omega_0 k)$ с выводом графиков вещественной и мнимой частей на интервале времени $n \in [0, N-1]$. Записать данный сигнал в виде комбинации двух вещественных последовательностей.


5. Вывести графики последовательностей $\delta_d (k), \sigma_d (k)$ и $s_1 (k)$, задержанных на m отсчетов, на интервале времени $n \in [0, N-1]$. Записать формулы задержанных последовательностей.


6. Сделать выводы.

## Выполнение работы

Пункты работы выполняется с помощью кода.

In [2]:
import math
import plotly.express as px
import numpy as np

#---------- Инициализация констант -------------
Nb = 13 # номер бригады
N = 30 + Nb % 5
T = 0.0005 * (1 + Nb % 3)
a = (-1) ** Nb * (0.8 + 0.005 * (Nb % 20))
C = 1 + Nb % 5
w0 = math.pi / (6 + Nb % 5)
m = 5 + Nb % 5

x_nT = np.arange(0, N*T, T) # делает до (N-1)*T
x_n = np.arange(0, N, 1) # делает до N-1

### 1. Смоделировать единичный цифровой импульс $\delta_d(k)$ с выводом графиков на интервале дискретного времени $nT \in [0, (N-1)T]$ и дискретного нормированного времени $n \in [0, N-1]$

In [3]:
def delta_d(k):
  if k == 0 :
    return 1
  else:
    return 0

vdelta_d = np.vectorize(delta_d)

print(vdelta_d)

px.scatter(x=x_nT, y=vdelta_d(x_nT), title="На интервале дискретного времени nT", height = 500).show()
px.scatter(x=x_n, y=vdelta_d(x_n), title="На интервале дискретного нормированного времени n", height = 500).show()



* __Взаимосвязь между дискретным и дискретным нормированным временем.__
  * Значения $𝑛𝑇$, где $𝑇 =\frac{1}{𝑓_д}$ – период дискретизации.
  * Значение $n$: $𝑛 = \frac{𝑛𝑇}{𝑇}$
* __Различие между цифровым единичным импульсом и дельта-функцией__.
  * Цифровой единичный импульс. Это аналог дельта-функции для дискретных систем, но в отличие него, - физически реализуемый сигнал.
  * Дельта-функция (она же функция Дирака), используется в аналоговых системах.

### 2. Смоделировать дискретный единичный скачок $\sigma_d(k)$ с выводом графиков на интервале дискретного времени $nT \in [0, (N-1)T]$ и дискретного нормированного времени $n \in [0, N-1]$

In [4]:
def sigma_d(k):
  if k >= 0 :
    return 1
  else:
    return 0

vsigma_d = np.vectorize(sigma_d)

px.scatter(x=x_nT, y=vsigma_d(x_nT), title="На интервале дискретного времени nT", height = 500).show()
px.scatter(x=x_n, y=vsigma_d(x_n), title="На интервале дискретного нормированного времени n", height = 500).show()

print("f_д =", 1/T, " Гц")

f_д = 1000.0  Гц


* __Соответствие между цифровым и аналоговым единичными скачками__
  * Функция единичного скачка (она же функция Хевисайда, или функция включения),
используется в аналоговых системах.
* __Чему равна частота дискретизации цифрового единичного скачка__
$$f_д = \frac{1}{T} = 1000 ~~Гц$$

### 3. Смоделировать дискретную экспоненциальную функцию $s_1(k)$ с выводом графиков на интервале дискретного времени $nT \in [0, (N-1)T]$ и дискретного нормированного времени $n \in [0, N-1]$. Пояснить соответствие между дискретной и аналоговой экспонентами.

In [5]:
def s_1(k, a):
  if k < 0: return 0
  return a ** k

s_1vec = np.vectorize(s_1)
values = s_1vec(x_n, a)

px.line(x=x_nT, y=values, title="На интервале дискретного времени nT", height = 500).show()
px.line(x=x_n, y=values, title="На интервале дискретного нормированного времени n", height = 500).show()


**Соответствие между дискретной и аналоговой экспонентами**.
Вид дискретной экспоненты определяется величиной и знаком параметра $a$.

### 4. Смоделировать дискретный комплексный гармонический сигнал  $s_2 ( k ) = C \exp ( j ω_0 k )$  с выводом графиков вещественной и мнимой частей на интервале времени  $n \in [0, N-1]$ . Записать данный сигнал в виде комбинации двух вещественных последовательностей.

In [6]:
def s2_re(k): #cos
  return C * math.cos(w0*k)

def s2_im(k): #sin
  return C * math.sin(w0*k)

vs2_re = np.vectorize(s2_re)
vs2_im = np.vectorize(s2_im)



px.scatter(x=x_n, y=vs2_re(x_n), title="Графиков вещественной части", height = 500).show()
px.scatter(x=x_n, y=vs2_im(x_n), title="Графиков мнимой части", height = 500).show()

print("C =", C)
print("w0 =", w0)

C = 4
w0 = 0.3490658503988659


__Сигнал в виде комбинации двух вещественных последовательностей:__
 $$s_2(k) = 4\cos(0.349 k) +j~4\sin(0.349 k)$$

### 5. Вывести графики последовательностей $δ_d(k)$, $σ_d(k)$ и $s_1(k)$, задержанных на $m$ отсчетов на интервале времени $n \in [0, N-1]$. Записать формулы задержанных последовательностей.

In [8]:
x_n_m = x_n - m

delta_d_values = np.vectorize(delta_d)(x_n_m)
sigma_d_values = np.vectorize(sigma_d)(x_n_m)
s_1_values = np.vectorize(s_1, otypes=[float])(x_n_m, a)

px.scatter(x=x_n, y=delta_d_values, title="delta_d", height = 350).show()
px.scatter(x=x_n, y=sigma_d_values, title="sigma_d", height = 350).show()
px.line(x=x_n, y=s_1_values, title="s_1", height = 350).show()



**Формулы задержанных последовательностей.**

__Единичный цифровой импульс__:
$$\delta_d(k) = \begin{cases}
1,~~ k=m \\
0,~~ k\neq 0
\end{cases} , ~~~~~
k= ~...,-2,-1,0,1,2,...$$

\\
__Дискретный единичный скачок__:
$$\sigma_d(k) = \begin{cases}
1,~~ k \geq m \\
0,~~ k < 0
\end{cases} , ~~~~~
k= ~...,-2,-1,0,1,2,...$$

\\
__Дискретная экспоненциальная функция:__:
$$
s_1(k) = \begin{cases}
a^{(k-m)},~~ k \geq m \\
0,~~ k < m
\end{cases} , ~~~~~
k= ~...,-2,-1,0,1,2,...$$


## Выводы

В ходе выполнение лабораторной работы были изучены функции: единичный цифровой импульс, единичный скачок, дискретная экспоненциальная функция и дискретный комплексный гармонический сигнал. Для первых трех функций были построены графики зависимости от дискретного и дискретного нормированного времени, а также проанализирована взаимосвязь между ними. Кроме того, для функций единичного цифрового импульса, единичного скачка, дискретной экспоненциальной функции были построены графики, задержанные на m отсчетов, и записаны их формулы. Для комплексного гармонического сигнала были построены графики вещественной и комплексной части. Таким образом, в ходе выполнения лабораторной работы были выработаны базовые навыки работы с сигналами: нормировка времени, отображение графиков, рассчет частоты сигнала, работа с комплексной частью сигнала.